src/host/multiple_ledger.h

```diff
@@ -0,0 +1,815 @@
+      {
+        if (f->get_start_idx() <= idx && idx <= f->get_last_idx())
+        {
+          return f;
+        }
+      }
+
+      // If the file is not in the cache, find the file from the ledger
+      // directory
+      std::map<size_t, std::string> all_files;
+      for (auto const& f : fs::directory_iterator(ledger_dir))
+      {
+        all_files.emplace(
+          get_start_idx_from_file_name(f.path().filename()),
+          f.path().filename());
+      }
+
+      auto f_ = std::upper_bound(
+        all_files.rbegin(),
+        all_files.rend(),
+        idx,
+        [](size_t idx, const std::pair<size_t, std::string>& v) {
+          return (idx >= v.first);
+        });
+
+      if (f_ == all_files.rend())
+      {
+        LOG_FAIL_FMT("Could not find ledger file on disk for idx {}", idx);
+        return nullptr;
+      }
+
+      // Emplace file in the max-sized read cache
+      auto match_file = std::make_shared<LedgerFile>(ledger_dir, f_->second);
+      if (files_read_cache.size() >= max_read_cache_size)
+      {
+        files_read_cache.erase(files_read_cache.begin());
+      }
+      files_read_cache.emplace_back(match_file);
+
+      return match_file;
+    }
+
+    std::shared_ptr<LedgerFile> get_file_from_idx(size_t idx)
+    {
+      if (idx == 0)
+      {
+        return nullptr;
+      }
+
+      // First, check if the file is in the list of files open for writing
+      auto f = std::upper_bound(
+        files.rbegin(),
+        files.rend(),
+        idx,
+        [](size_t idx, const std::shared_ptr<LedgerFile>& f) {
+          return idx >= f->get_start_idx();
+        });
+
+      if (f != files.rend())
+      {
+        return *f;
+      }
+
+      // Otherwise, return file from read cache
+      return get_file_from_cache(idx);
+    }
+
+    std::shared_ptr<LedgerFile> get_latest_file() const
+    {
+      if (files.empty())
+      {
+        return nullptr;
+      }
+      return *(files.rbegin()++);
+    }
+
+  public:
+    MultipleLedger(
+      const std::string& ledger_dir,
+      ringbuffer::AbstractWriterFactory& writer_factory,
+      size_t chunk_threshold,
+      size_t max_read_cache_size = max_read_cache_size_default) :
+      ledger_dir(ledger_dir),
+      chunk_threshold(chunk_threshold),
+      to_enclave(writer_factory.create_writer_to_inside()),
+      max_read_cache_size(max_read_cache_size)
+    {
+      if (chunk_threshold == 0 || chunk_threshold > max_chunk_threshold_size)
+      {
+        throw std::logic_error(fmt::format(
+          "Error: Ledger chunk threshold should be between 1-{}",
+          max_chunk_threshold_size));
+      }
+
+      if (fs::is_directory(ledger_dir))
+      {
+        // If the ledger directory exists, recover ledger files from it
+        for (auto const& f : fs::directory_iterator(ledger_dir))
+        {
+          files.push_back(
+            std::make_shared<LedgerFile>(ledger_dir, f.path().filename()));
```

Comment:
Looks like `LedgerFile` always reconstructs the full path, so we don't need to split these? (We can construct `LedgerFile` from just `f.path()`?)

Reply:
I think this makes the two constructors of `LedgerFile` cleaner: 
- `LedgerFile(dir, start_idx)` that create a fresh new ledger file from `start_idx`
- `LedgerFile(dir, file_name)` that restores an existing ledger file 

We need to get the `directory` in `LedgerFile` so we'd have to split it in there anyway.